### PR TITLE
NEP 22: Contract Update Standard

### PR DESCRIPTION
**Pull request by superboyiii**
Close: https://github.com/neo-project/proposals/issues/135

### CODE REVIEW

File: nep-xx.mediawiki
Comment (ixje) [inline nep-xx.mediawiki:52]:
```suggestion
|                                                              |      | This is an optional function which will be automatically executed when a contract is first deployed or updated. |
```

File: nep-xx.mediawiki
Comment (ixje) [inline nep-xx.mediawiki:56]: > The function SHOULD check whether <code>update</code> returns true. If not, continue execution. If true, return.

I don't think this is correct advice. It seems perfectly reasonable to me to do something like this (pseudo)

```
deploy(data, update) {
   if !update {
      // initial deploy, put default values into storage
   } else {
      // reset part of storage to default
   }
}
```

File: nep-xx.mediawiki
Comment (ixje) [inline nep-xx.mediawiki:82]: Why suggest the requirement that both manifest and NEF must be present? The management contract requires one of the two or both.
https://github.com/neo-project/neo/blob/5f4c14e3801c691ceccb5705c634e9e55dd3bba1/src/Neo/SmartContract/Native/ContractManagement.cs#L233

File: nep-xx.mediawiki
Comment (superboyiii) [reply nep-xx.mediawiki:56]: In fact, either `initial deploy` or `update` will send notification, no need to waste storage.
https://github.com/neo-project/neo/blob/5f4c14e3801c691ceccb5705c634e9e55dd3bba1/src/Neo/SmartContract/Native/ContractManagement.cs#L99

Comment (ixje) [reply nep-xx.mediawiki:56]: I'm aware of the existing notification, but that's not the issue I'm trying to point out. The way I'm understanding the text is that it suggests to do the following
```
_deploy(data, update) {
   if update { return }
   // else continue execution of remaining statements
}
```
if that's the case then the whole function should not have a need to have an `update` parameter to begin with. Since it does have this param, it is reasonable that a contract might do something different when updated vs when first deployed. The suggestion would therefore be to remove this text
> The function SHOULD check whether update returns true. If not, continue execution. If true, return.

Comment (superboyiii) [reply nep-xx.mediawiki]: Done

File: nep-xx.mediawiki
Comment (superboyiii) [reply nep-xx.mediawiki]: Done

**Review by ixje — DISMISSED** [review]
LGTM

File: nep-xx.mediawiki
Comment (roman-khimov) [inline nep-xx.mediawiki:22]: I'm not entirely sure we should put `_deploy` there on the same line as `update` or `destroy` because it's a system-defined method, not a convention (it's even briefly mentioned on https://docs.neo.org/docs/en-us/reference/scapi/framework/native/ContractManagement/Deploy.html). But maybe this NEP will make it somewhat more visible, probably not all developers even know about it.

File: nep-xx.mediawiki
Comment (roman-khimov) [inline nep-xx.mediawiki:50]: `...executed by ContractManagement contract...`?

File: nep-xx.mediawiki
Comment (roman-khimov) [inline nep-xx.mediawiki:52]: `contract-specific purpose`? Mention that it's the same `data` that is accepted by ContractManagement's Deploy and Update?

File: nep-xx.mediawiki
Comment (roman-khimov) [inline nep-xx.mediawiki:54]: I don't think `update` returns anything, it's just a `Boolean`, maybe `update is true when contract is updated (via Update) and it's false on initial deployment (via Deploy)`.

Comment (roman-khimov) [reply nep-xx.mediawiki:54]: And BTW, maybe it's important to tell that when contract is updated it's the _new_ contract's `_deploy` that is called, not old one.

File: nep-xx.mediawiki
Comment (roman-khimov) [inline nep-xx.mediawiki:82]: Mention that it will be passed to `_deploy`?

File: nep-xx.mediawiki
Comment (roman-khimov) [inline nep-xx.mediawiki:84]: BTW, we don't have any "ownability" NEP that will define what is "owner" and how to deal with this concept, but it's probably not critical for this NEP.

**Review by roman-khimov — COMMENTED** [review]
It'd be nice to have this, I'd change our examples to use this standard interface.

File: nep-xx.mediawiki
Comment (superboyiii) [reply nep-xx.mediawiki:84]: Done

File: nep-xx.mediawiki
Comment (superboyiii) [reply nep-xx.mediawiki]: Done

File: nep-xx.mediawiki
Comment (superboyiii) [reply nep-xx.mediawiki]: Done

File: nep-xx.mediawiki
Comment (superboyiii) [reply nep-xx.mediawiki]: Done

File: nep-xx.mediawiki
Comment (superboyiii) [reply nep-xx.mediawiki]: Done

File: nep-xx.mediawiki
Comment (superboyiii) [reply nep-xx.mediawiki]: Done

**Review by roman-khimov — APPROVED** [review]

File: nep-xx.mediawiki
Comment (shargon) [inline nep-xx.mediawiki:58]: In the view https://github.com/neo-project/proposals/pull/154/files?short_path=6caa849#diff-6caa8493f36ba42782e59054f66c45a43f6c170927ce9383eccda13535890ea1 this region is wrong displayed

Comment (superboyiii) [reply nep-xx.mediawiki:58]: Fixed.

**Comment by erikzhang** [comment]
Add `_initialize`?

**Comment by roman-khimov** [comment]
`_initialize` is something different to me, not directly related to deploy/update. Maybe it's in some other document (but I wasn't able to find it on https://docs.neo.org/) or other NEP.

**Comment by superboyiii** [comment]
@roman-khimov @erikzhang Added. Please review again.

**Comment by roman-khimov** [comment]
NEP-22:

> provide systems with a generalized interaction mechanism for smart contract initial deploy, update and destroy. 

I don't see `_initialize` being a part of this NEP.

**Comment by erikzhang** [comment]
> I don't see _initialize being a part of this NEP.

I'm not sure if it should be part of this PR, but obviously it should be included in a document.

Also, don't forget the `verify` method.

**Comment by superboyiii** [comment]
> > I don't see _initialize being a part of this NEP.
> 
> I'm not sure if it should be part of this PR, but obviously it should be included in a document.
> 
> Also, don't forget the `verify` method.

Added.

**Comment by superboyiii** [comment]
@erikzhang @roman-khimov Is it OK?

**Comment by roman-khimov** [comment]
It's a method soup now to me. I'd rather have separate NEPs for `_initialize`, `_deploy` and `verify` (three NEPs, seems to be perfectly fine to me, they are different unrelated things) and one for `update`/`destroy`. I want to be able to specify "NEP-XYZ" in `SupportedStandards` and expect all the tooling around to understand that this contract has standardized `update`/`destroy` method pair that can be interacted with. "NEP-22" doesn't give this meaning right now.

File: nep-xx.mediawiki
Comment (roman-khimov) [reply nep-xx.mediawiki]: Related to #159, btw.

**Comment by ixje** [comment]
> `_initialize` is something different to me, not directly related to deploy/update. Maybe it's in some other document (but I wasn't able to find it on https://docs.neo.org/) or other NEP.

I can kind of see the relationship. `update` can call `_deploy` which in turn can call `_initialize` again. Ideally there would be a flow diagram included to make it clear. `verify` on the other hand seems totally unrelated and worthy of a separate document if we're splitting anyway

**Comment by roman-khimov** [comment]
> I can kind of see the relationship

Is it related? Probably so. Can be it be used _without_ all the other things? Sure it is. Can all the other things in this proposal be used without `_initialize`? Sure they are. We can have as many standards as we need and they can be related, but different standards for different things. `NEP-XYZ` in `SupportedStandards` should have some specific meaning. And `SupportedStandards` can have as many `NEP-X, NEP-Y, NEP-Z` as needed.

**Comment by EdgeDLT** [comment]
> And SupportedStandards can have as many NEP-X, NEP-Y, NEP-Z as needed.

I agree in principle with the idea of having independent standards. But I suspect there are many others like myself who also have a mental block at the future thought of these long lists of numbers with no meaning. Maybe we could package them together under a standard group, but at that point, is it that much different from what we have now?

**Comment by cschuchardt88** [comment]
> > I can kind of see the relationship
> 
> Is it related? Probably so. Can be it be used _without_ all the other things? Sure it is. Can all the other things in this proposal be used without `_initialize`? Sure they are. We can have as many standards as we need and they can be related, but different standards for different things. `NEP-XYZ` in `SupportedStandards` should have some specific meaning. And `SupportedStandards` can have as many `NEP-X, NEP-Y, NEP-Z` as needed.

_initialize is used when verifying a witness and/or in invocationScripts & verificationScript.  It gets initialized before invoking those scripts. I don't know if there is a standard for that function may need to create one. I know neo uses it internal for the above mentioned. also anytime contract is invoked. I do believe the attribute "InitializeValue" creates this function, don't recall. Would have to check

File: nep-22.mediawiki
Comment (roman-khimov) [inline nep-22.mediawiki:91]: How come it's a `String` btw? ContractManagement itself expects a `ByteArray`.

Comment (cschuchardt88) [reply nep-22.mediawiki:91]: It doesn't matter for arguments. In the vm its already byte array; to be specific `ByteString`; can be anything for the most part. It all get compiled down to byte array.

Comment (roman-khimov) [reply nep-22.mediawiki:91]: That's not the point, these are different things as per NEP-14 and one can expect some type symmetry between `ContractManagement` definition of `update` and this one. Manifests matter, especially if you try to generate some code using them.

Comment (cschuchardt88) [reply nep-22.mediawiki:91]: Then why not make a `ByteString`?

Comment (superboyiii) [reply nep-22.mediawiki]: > That's not the point, these are different things as per NEP-14 and one can expect some type symmetry between `ContractManagement` definition of `update` and this one. Manifests matter, especially if you try to generate some code using them.

Fixed

**Comment by roman-khimov** [comment]
My take on it is still the same, three NEPs here would be much better than a single one:

> It's a method soup now to me. I'd rather have separate NEPs for `_initialize`, `_deploy` and `verify` (three NEPs, seems to be perfectly fine to me, they are different unrelated things) and one for `update`/`destroy`. I want to be able to specify "NEP-XYZ" in `SupportedStandards` and expect all the tooling around to understand that this contract has standardized `update`/`destroy` method pair that can be interacted with. "NEP-22" doesn't give this meaning right now.

This one was supposed to be about update/destroy initially.

File: nep-22.mediawiki
Comment (shargon) [inline nep-22.mediawiki:67]: In fact we accept verify with args too

File: nep-22.mediawiki
Comment (shargon) [inline nep-22.mediawiki:12]: and verify 
```suggestion
This proposal outlines a basic method guideline for the NEO blockchain that will provide systems with a generalized interaction mechanism for smart contract initial deploy, update, verify and destroy.
```

**Comment by superboyiii** [comment]
@shargon done

**Review by roman-khimov — CHANGES_REQUESTED** [review]
https://github.com/neo-project/proposals/pull/154#issuecomment-1988307672

I can't imagine how to use this NEP in its current form. It needs a split.

**Comment by Jim8y** [comment]
> [#154 (comment)](https://github.com/neo-project/proposals/pull/154#issuecomment-1988307672)
> 
> I can't imagine how to use this NEP in its current form. It needs a split.

@roman-khimov this one can be the most general description of base methods, as for specific method that should have a seperate NEP, we can create another NEPs for them, like Verify.

**Comment by roman-khimov** [comment]
Let's get back to #135 that started all of this for a moment. What people really wanted there is a _convention_ for methods that contract authors can use to have some common way to update/destroy contracts. In this sense it's somewhat similar to NEP-11/17/24. And this is only about two methods: `update` and `destroy`, so that I wouldn't want to do
```
func obnovlyashka(data Array)
```
to update my contract (because 3rd-party tools couldn't easily create an update transaction for me in this case, unlike when they deal with a compliant `update`).

But what we also have here now are:
 * `verify`, which has nothing to do with update/destroy, it's just a different mechanism
 * `_deploy`, which _is_ related to `update`, but is still somewhat different in that it can be omitted completely, it can be implemented in a contract that doesn't have `update` and, most importantly, it's not something you can call externally yourself, it's a native contract mechanism

At least we need to move `verify` out. It'd be perfect to have a NEP for it, but it's a separate matter. I would also move `_deploy` out since it has a broader definition (not just updates, but deployments as well, works irrespective of how your exposed-from-contract update methods are called), then this NEP could just refer to it.

**Comment by AnnaShaleva** [comment]
Totally agree with Roman, we should update this NEP to standardize a single update method. The rest of methods standards should be moved to a separate granular NEPs.

**Comment by superboyiii** [comment]
@roman-khimov @shargon @cschuchardt88 @Jim8y Now I changed it to `Contract Update Standard`. I will submit `destroy standard` as well.

File: nep-22.mediawiki
Comment (shargon) [inline nep-22.mediawiki:49]:
```suggestion
This method will be automatically executed by <code>ContractManagement</code> contract when a contract is updated. When <code>update</code> in <code>_deploy</code> is true, ContractManagement contract will update new contract state by this method.
```

File: nep-22.mediawiki
Comment (shargon) [inline nep-22.mediawiki:51]: In fact, you can send one of them as NULL, but not both

https://github.com/neo-project/neo/blob/978828005474dce455fe368e6e275f30293a5411/src/Neo/SmartContract/Native/ContractManagement.cs#L274

you can update only the manifest

File: nep-22.mediawiki
Comment (roman-khimov) [inline nep-22.mediawiki:17]: It's more like we didn't have an agreement on how deployed contracts expose this functionality to their admins. Because ContractManagement itself doesn't care about code that calls it.

File: nep-22.mediawiki
Comment (roman-khimov) [inline nep-22.mediawiki:21]: Reference NEP-29 for details here.

File: nep-22.mediawiki
Comment (roman-khimov) [inline nep-22.mediawiki:49]: That's not how it works. It's an externally (by transaction) called method that calls ContractManagement internally. Then it does the magic.

File: nep-22.mediawiki
Comment (roman-khimov) [inline nep-22.mediawiki:55]: SHOULD check, this can be done (or not done) in various ways.

**Comment by superboyiii** [comment]
@shargon @roman-khimov  Done.

File: nep-22.mediawiki
Comment (vncoelho) [inline nep-22.mediawiki:49]: will will

File: nep-22.mediawiki
Comment (vncoelho) [inline nep-22.mediawiki:53]: "For example some kinds of json parameters can be put into data for specific check."
Some examples here?

File: nep-22.mediawiki
Comment (vncoelho) [inline nep-22.mediawiki:55]: This function? Which function?

File: nep-22.mediawiki
Comment (roman-khimov) [inline nep-22.mediawiki:17]: `s/NEO/Neo`. `s/standize/a standard/`? I'm also not sure we need to say anything about callbacks. It's about standard API to update contracts, what it does internally is not motivation probably.

File: nep-22.mediawiki
Comment (roman-khimov) [inline nep-22.mediawiki:21]: I don't think it's a part of specification. We're specifying something different here.

File: nep-22.mediawiki
Comment (roman-khimov) [reply nep-22.mediawiki:49]: Probably just `MUST call`

File: nep-22.mediawiki
Comment (roman-khimov) [inline nep-22.mediawiki:51]: Maybe refer to NEP-16 and NEP-15 here. I also don't understand 

> It will be passed to <code>_deploy</code>.

here. Only `data` is going to be passed and that's exactly where a reference to NEP-29 is needed.

File: nep-22.mediawiki
Comment (roman-khimov) [reply nep-22.mediawiki:55]: Maybe better as `this method`, since we're talking about `update` _method_ here. Also, I don't understand what is to `check whether the <code>signer</code> address equals the <code>owner</code> hash of contract`. Transaction can have a lot of signers. We should care more about owner allowing this action (`CheckWitness`). And NEP-30 is a different thing in general, _most_ of contracts will not have contract-based owners, they're usually just multisig.

**Comment by superboyiii** [comment]
All done @roman-khimov @vncoelho

File: nep-22.mediawiki
Comment (vncoelho) [inline nep-22.mediawiki:21]: Maybe explain current consequences if not implemented in the first version.

File: nep-22.mediawiki
Comment (vncoelho) [inline nep-22.mediawiki:49]: exectued

File: nep-22.mediawiki
Comment (vncoelho) [inline nep-22.mediawiki:51]: I think maybe more details on the consequences here. I mean, the differences that can happen if updated in two calls.

File: nep-22.mediawiki
Comment (Jim8y) [reply nep-22.mediawiki:49]: executed？

File: nep-22.mediawiki
Comment (superboyiii) [reply nep-22.mediawiki]: We can guide them to read the relevant NEP standards listed below.

File: nep-22.mediawiki
Comment (superboyiii) [reply nep-22.mediawiki]: Done

File: nep-22.mediawiki
Comment (superboyiii) [reply nep-22.mediawiki]: Done

File: nep-22.mediawiki
Comment (vncoelho) [inline nep-22.mediawiki:21]: good

**Review by vncoelho — DISMISSED** [review]

File: nep-22.mediawiki
Comment (roman-khimov) [reply nep-22.mediawiki]: It can be implemented in a different way, I mean non-NEP-22 method that still does the update. Also, the way it's worded now is like "if you have a bug, please implement this method", while in fact it's too late if you have a bug and no way to update the contract.

**Comment by superboyiii** [comment]
@roman-khimov Done

File: nep-22.mediawiki
Comment (roman-khimov) [inline nep-22.mediawiki:49]: s/CALL/call/

File: nep-22.mediawiki
Comment (roman-khimov) [inline nep-22.mediawiki:70]: I think we don't need a space in this list, although it renders fine in the end.

File: nep-22.mediawiki
Comment (roman-khimov) [inline nep-22.mediawiki:51]: Unfortunately, it's not markdown, see how it renders with "view file" on GH, links need to be fixed (see https://github.com/neo-project/proposals/blob/a88f56086f8ca607af2c0be3c0d87dec2b3ca4ca/nep-25.mediawiki?plain=1#L252 and other NEPs).

File: nep-22.mediawiki
Comment (roman-khimov) [inline nep-22.mediawiki:21]: Maybe `... implement this NEP to be able to fix bugs or otherwise improve functionality`. Just to not be misinterpreted for "implement after you have a bug", no, it's "think about it before you have bugs".

**Review by roman-khimov — COMMENTED** [review]
Editorial nitpicks, but OK otherwise.

File: nep-22.mediawiki
Comment (superboyiii) [reply nep-22.mediawiki:70]: C# example is necessary as well as GO.

**Comment by superboyiii** [comment]
@roman-khimov Done

**Review by roman-khimov — APPROVED** [review]

**Review by Jim8y — APPROVED** [review]

File: nep-22.mediawiki
Comment (AnnaShaleva) [inline nep-22.mediawiki:6]: This standard is accepted, shall we update the README.md?

File: nep-22.mediawiki
Comment (AnnaShaleva) [inline nep-22.mediawiki:13]:
```suggestion
This proposal outlines a standard for the Neo blockchain that will provide systems with a generalized interaction mechanism for smart contract update.
```

**Review by AnnaShaleva — APPROVED** [review]

**Review by AnnaShaleva — APPROVED** [review]
Sorry for this minor fix, I couldn't resist, and now all previous approvals are gone.  @Jim8y, @roman-khimov, need your approvals one more time.

File: nep-22.mediawiki
Comment (roman-khimov) [reply nep-22.mediawiki:6]: I think we better merge this, #182 and then update the README, because there are some omissions there anyway.

**Review by roman-khimov — APPROVED** [review]

File: nep-22.mediawiki
Comment (roman-khimov) [reply nep-22.mediawiki:6]: The scheme is exactly to avoid endless reapprovals, the document is fine now and that's the main thing here.

**Comment by Jim8y** [comment]
> Sorry for this minor fix, I couldn't resist, and now all previous approvals are gone. @Jim8y, @roman-khimov, need your approvals one more time.

That is totally ok to comment on minor fixes, you are the strongest sheld we have to ensure eveything works properly, your concern definately will be addressed after the merge. Just approve an NEP takes too long while fixing minor issues can be in another track, along with file name/nep number/readme.

**Pull request merged**
